### PR TITLE
Add Jupyter application

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Jupyter is a suite of tools widely used in data science, scientific computing and machine learning.

## Proposal

- [x] add the jupyter application

### CODE REVIEW

File: apps/jupyter/templates/services/app/dc.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      containers:
+        - name: jupyter
+          image: "{{ jupyter_image_name }}:{{ jupyter_image_tag }}"
+          imagePullPolicy: Always
+          envFrom:
+            - secretRef:
+                name: "{{ jupyter_secret_name }}"
+          env:
+            - name: JUPYTER_ENABLE_LAB
+              value: "yes"
```
Comment (lunika) [inline apps/jupyter/templates/services/app/dc.yml.j2:47]: Why not configuring this variable ?

File: apps/jupyter/templates/services/app/dc.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+          volumeMounts:
+            - name: jupyter-v-home
+              mountPath: /home/jovyan
```
Comment (lunika) [inline apps/jupyter/templates/services/app/dc.yml.j2:50]: Can you add a comment explaining why this particular path ? I suppose this home dir is created in the image used but a comment will be better here IMO

File: apps/jupyter/templates/services/app/route.yml.j2 [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  labels:
+    app: jupyter
+    service: app
+    version: "{{ jupyter_image_tag }}"
+  annotations:
+    haproxy.router.openshift.io/disable_cookies: "true"
+    haproxy.router.openshift.io/balance: "roundrobin"
+{% if jupyter_app_ip_whitelist | length %}
+{% for ip in jupyter_app_ip_whitelist %}
+    haproxy.router.openshift.io/ip_whitelist: "{{ ip }}"
```
Comment (lunika) [inline apps/jupyter/templates/services/app/route.yml.j2:15]: Are you sure this is working ? Can you multiple time the same annotations ? Reading the documentation I'm not sure : https://docs.openshift.com/dedicated/3/architecture/networking/routes.html#whitelist
Does it work behind a proxy ?

File: apps/jupyter/templates/services/app/dc.yml.j2 (hunk 1 above)
Comment (jmaupetit) [reply apps/jupyter/templates/services/app/dc.yml.j2:47]: We could!

File: apps/jupyter/templates/services/app/dc.yml.j2 (hunk 2 above)
Comment (jmaupetit) [reply apps/jupyter/templates/services/app/dc.yml.j2:50]: This home directory is required to store files created in working sessions (notebooks, data, etc.).

File: apps/jupyter/templates/services/app/route.yml.j2 (hunk 3 above)
Comment (jmaupetit) [reply apps/jupyter/templates/services/app/route.yml.j2:15]: My implementation was broken, thanks! And Yes: it works behind a proxy.

**Comment by jmaupetit** [comment]
I've added fixes for your comments @lunika :pray:

**Review by lunika — APPROVED** [review]

**Pull request merged**